sympy/concrete/zeilberger.py

```
@@ -0,0 +1,144 @@
+    function G of n and k such that G / F is rational and
+
+    a_0F(n, k) + a_1F(n + 1, k) + ... + a_JF(n + j, k) = G(n, k + 1) - G(n, k).
+
+    If such things exists this function returns a pair consisting of
+    [a_0, a_1, ... , a_J] and the function G / F, otherwise it returns
+    'try higher order', suggesting you try to find a higher order recurrence,
+    i.e. increase J. If F is 'proper hypergeometric' (see page 64 of 'A = B'
+    for a definition) then such a recurrence always exists for some J.
+
+    The main application of this is to sum k over some suitable values so that
+    the G telescopes and you obtain a recurrence for the sum of F(n, k).
+
+    Examples
+    ========
+    Helping us discover that sum of binomial(n - k - 1, k) over k is fibonacci
+
+    >>> from sympy.concrete.zeilberger import zb_recur
+    >>> from sympy.abc import n, k, x
+    >>> from sympy import binomial
+    >>> F = binomial(n - k - 1, k)
+    >>> zb_recur(F, 1, n, k)
+    'try higher order'
+    >>> zb_recur(F, 2, n, k)
+    ([-1, -1, 1], k*(k - n)/((2*k - n)*(2*k - n - 1)))
+
+    Discovering a recurrence for binomial(n, k)**3, the resulting recurrence for
+    the sum can't be solved nicely, we don't care to look at G.
+
+    >>> F = binomial(n, k)**3
+    >>> zb_recur(F, 2, n, k)[0]
+    [-2, -(7*n**2 + 21*n + 16)/(4*(n + 1)**2), (n + 2)**2/(4*(n + 1)**2)]
+
+    F may have other symbols than n and k, here we discover the sum from k = 0 to n to be
+    x / (x + n).
+
+    >>> F = (-1)**k * binomial(n, k) / binomial(x + k, k)
+    >>> zb_recur(F, 1, n, k)
+    ([-n - x, n + x + 1], k*(k + x)/(k - n - 1))
+    """
+    a, w = symbols('a, w', cls = Function)
+
+    p, p_2, p_3 = _zb_gosper(F, J, n, k)
+    # This provides us the polynomials appearing in (6.3.11) of 'A = B', p_3
+    # here is p_3(k-1)
+
+    b_deg = _find_b_deg(p_2, p_3, p)
+
+    if b_deg < 0:
+        return "try higher order"
```

Functions should not return strings. If there is an error, it should raise an exception (like `raise ValueError("try higher order")`).